lib/instrument.ex

```diff
@@ -0,0 +1,390 @@
+    fun = instrument_expr(instrumenter, f)
+    args = map_expr(as, instrumenter)
+    case instrumenter.is_instrumentable_function(m, f) do
+      true -> instrumenter.handle_function_call({:call, line, {:remote, line2, module, fun}, args})
+      _ -> {:call, line, {:remote, line2, module, fun}, args}
+    end
+  end
+  def instrument_function_call(instrumenter, {:call, line, f, as}) do
+    fun = instrument_expr(instrumenter, f)
+    args = map_expr(as, instrumenter)
+    {:call, line, fun, args}
+  end
+
+  @doc "The receive might be handled differently, therefore it has its own function"
+  def instrument_receive(instrumenter, {:receive, line, cs}) do
+    {:receive, line, map(cs, instrumenter, &instrument_clause/2)}
+  end
+  def instrument_receive(instrumenter, {:receive, line, cs, e, b})  do
+    {:receive, line, map(cs, instrumenter, &instrument_clause/2),
+      instrument_expr(instrumenter, e), map_expr(b, instrumenter)}
+  end
+
+  @doc """
+  Prepends the call to `to_be_wrapped_call` by a call to `new_call`.
+  The result of `new_call` is ignored.
+  """
+  def prepend_call(to_be_wrapped_call, new_call) do
+    {:block, [generated: true], [new_call, to_be_wrapped_call]}
+  end
+
+  @doc """
+  Enocdes a call given as tuple `{m, f, a}` as Elixir values into an
+  abstract erlang form
+  """
+  def encode_call({m, f, a}) do
+    line = 0 # [generated: true]
+    {:call, line,
+      {:remote, line,
+        {:atom, line, m},
+        {:atom, line, f}},
+      Enum.map(a, &encode_value/1)}
+  end
+  def encode_call(m, f, a) when is_atom(m) and is_atom(f) and is_list(a),
+    do: encode_call({m, f, a})
+
+  @doc "Encodes a value"
+  def encode_value(nil), do: {:nil, [generated: true]}
+  def encode_value(value) when is_atom(value), do: {:atom, [generated: true], value}
+  def encode_value(value) when is_integer(value), do: {:integer, [generated: true], value}
+  def encode_value(value) when is_float(value), do: {:float, [generated: true], value}
+  def encode_value(value) when is_binary(value), do:
+    {:bin, [generated: true],
+      [{:bin_element, [generated: true], {:string, [generated: true], String.to_charlist(value)},
+        :default, :default}]}
+  def encode_value([]), do: encode_value(nil)
+  def encode_value(l) when is_list(l) do
+    l
+    |> Enum.reverse()
+    |> Enum.reduce({nil, 0}, fn v, acc -> {:cons, 0, encode_value(v), acc} end)
+  end
+  def encode_value(t) when is_tuple(t) do
+    {:tuple, [generated: true],
+      t
+      |> Tuple.to_list()
+      |> Enum.map(&encode_value/1)}
+  end
+  def encode_value(_unknown), do: throw ArgumentError
+
+  @doc "Encodes a call to `:erlang.yield()"
+  def call_yield do
+    encode_call({:erlang, :yield, []})
+  end
+
+  # Generate the matcher for interestng functions by a macro
+  # This list is taken from the opensource version of Pulse
+  # (https://github.com/prowessproject/pulse-time/blob/master/src/pulse_time_scheduler.erl)
```

Comment:
Do we need to take care of the license here? PropCheck uses GPLv3, Pulse 3-BSD if I am not mistaken.